### PR TITLE
feat: plug NodeSecure runtime configuration

### PR DESCRIPTION
**Pull request by antoine-coulon**
Linked to #2 

This PR aims to integrate the **@nodesecure/rc** lib which would be another way to provide a runtime configuration in addition to the **API** and **CLI** options.

More generally it also introduces a way to manage different types of configuration in a transparent way for the @nodesecure/ci runner by providing standardization mechanisms towards a **Nsci Configuration**.

Consequently here are few semantic choices that can be discussed:

- **Nsci** configuration becomes the standard interface for a valid configuration in order to run the pipeline
- **External** configuration becomes the standard format for every external configuration (API, CLI, .nodesecurerc). Each configuration has to implement an _adapter_ than converts the input format of the configuration to the **external configuration format**.
- Once **CLI, API or .nodesecurerc** configuration is adapted to the **External** configuration, the **External** configuration can be _standardized_  and _validated_ to the **Nsci** format.

Take for example:

- **cli configuration**
```typescript
export type CliConfig = {
  directory: string;
  strategy: Nsci.InputStrategy;
  vulnerabilities: Nsci.Severity;
  warnings: Nsci.Warnings;
  reporters: string | Nsci.ReporterTarget[];
};
```

- **.nodesecurerc configuration, described differently**
```typescript
export interface RC {
  version: string;
   i18n?: i18n.languages;
   strategy?: vuln.Strategy.Kind;
   ci?: {
     reporters?: ("console" | "html")[];
     vulnerabilities?: {
        severity?: "medium" | "high" | "critical" | "all";
     };
     warnings?: CiWarnings | Record<jsxray.kindWithValue | "unsafe-import", CiWarnings>;
   };
}
```

These two different configurations can be defined by the user upstream the **pipeline runner** but we want to abstract this
complexity for the runner which should consume a standard format agnostic of configuration changes.

For that, I propose to describe a target standard format for each of these configurations such as:
```typescript
export type ExternalRuntimeConfiguration = {
  directory: string;
  strategy: Nsci.InputStrategy;
  vulnerabilities: Nsci.Severity;
  warnings: Nsci.Warnings;
  reporters: string | Nsci.ReporterTarget[];
};
```

Each of the config could then be converted to this format above by implementing a common adapter:
```typescript
export type ExternalConfigAdapter<T> = {
  adaptToExternalConfig: (config: T) => ExternalRuntimeConfiguration;
};
```

Once the config is adapted to the ExternalRuntimeConfiguration, its an easy work to sanitize/validate/complete the **Nsci** standard configuration (check **src/config/external/standardize.ts**)

### CODE REVIEW

**Review by tony-go — COMMENTED** [review]
👏 Awesome!

My only concern is about testing. I read a lot of code and only a few high-level tests.

Maybe we could merge it, then we could start an initiative to improve coverage... @antoine-coulon WDYT?

File: bin/cli.ts [hunk 1]
```diff
@@ -2,55 +2,62 @@
...
+
+program
+  .command("run")
   .option(
     "-d, --directory",
     "Root directory from which the analyses will be run.",
     process.cwd()
   )
-  .example("cli.js --directory=/Users/user1/myproject")
+  .example("cli.js run --directory=/Users/user1/myproject")
```
[the remark this answers is not in the document]
Comment (antoine-coulon) [reply bin/cli.ts:32]: Yes you're right, should be `nsci` as it is the .bin associated with the library.

File: src/analysis/interpretation/vulnerabilities.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+import { Nsci } from "../../configuration/standard/index.js";
+import { Maybe } from "../../types/index.js";
+
+import { fromBooleanToCheckResult, CheckableFunction } from "./checkable.js";
+
+function fromSeverityToNumber(
+  severity: Maybe<Strategy.Severity | "all">
+): number {
+  const severities = {
```
[the remark this answers is not in the document]
Comment (antoine-coulon) [reply src/analysis/interpretation/vulnerabilities.ts:11]: I'm not used to using this semantic for constants but let's make it uniform through the org.

File: src/analysis/interpretation/vulnerabilities.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    low: 1,
+    info: 0,
+    all: 0
+  };
+
+  if (severity !== undefined) {
+    return severities[severity];
+  }
+
+  const DEFAULT_SEVERITY = 0;
```
[the remark this answers is not in the document]
Comment (antoine-coulon) [reply src/analysis/interpretation/vulnerabilities.ts:24]: I'll make it uniform

File: src/analysis/interpretation/vulnerabilities.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+ * We must ensure that each vulnerability with equal or higher severity than
+ * the one defined in the runtime configuration is caught.
+ */
+function findAllVulnsExceedingSeverityThreshold(
+  vulnerabilities: Strategy.StandardVulnerability[],
+  severityThreshold: Strategy.Severity | "all"
+): Strategy.StandardVulnerability[] {
+  return vulnerabilities.filter((vuln) =>
+    isVulnExceedingSeverityThreshold(vuln.severity, severityThreshold)
+  );
```
[the remark this answers is not in the document]
Comment (antoine-coulon) [reply src/analysis/interpretation/vulnerabilities.ts:53]: I see your point, I'm probably going to implement an high order function to inject the `severityThreshold` before the filter callback method allowing us to be more declarative.

File: src/reporting/reporters/post-pipeline/console/global-warnings.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
+import { GlobalWarning } from "@nodesecure/scanner/types/scanner";
+import pluralize from "pluralize";
+
+import {
+  ConsoleMessage,
+  consolePrinter
+} from "../../../../../lib/console-printer/index.js";
```
[the remark this answers is not in the document]
Comment (antoine-coulon) [reply src/reporting/reporters/post-pipeline/console/global-warnings.ts:7]: I agree, the goal of the lib is to be implemented and maintained outside of the `@nodesecure/ci` scope, that's why I put it in the first place outside of the `src/` folder

File: .eslintrc [hunk 6]
```diff
@@ -27,6 +27,7 @@
   "rules": {
     "indent": "off",
     "brace-style": "off",
+    "id-length": "off",
```
[the remark this answers is not in the document]
Comment (tony-go) [reply .eslintrc:30]: I created: https://github.com/NodeSecure/eslint-config/issues/17

File: README.md [hunk 7]
```diff
@@ -112,27 +106,88 @@ Then run it
...
+[The documentation of the @nodesecure/ci GitHub Action is detailed here](https://github.com/marketplace/actions/nodesecure-continuous-integration)
 
 ### Custom configuration
 
-For now, the configuration is managed internally and is only configurable there:
+#### .nodesecurerc (runtime configuration file)
+
+A custom configuration can now be provided using the brand new **.nodesecurerc** config file.
+
+> **When a .nodesecurerc file is present in the project it will take priority over any other
```
[the remark this answers is not in the document]
Comment (tony-go) [reply README.md:130]: Didn't see your update on the PR.

[the remark this answers is not in the document]
Comment (antoine-coulon) [reply README.md:130]: @tony-go I added a small **Mixing configurations** sub section at the end of the **Custom configuration** section

[the remark this answers is not in the document]
Comment (tony-go) [reply README.md:130]: Got it ^^ Sorry

**Review by tony-go — APPROVED** [review]
LGTM

File: src/configuration/environment/index.ts [hunk 8]
```diff
@@ -1,20 +1,20 @@
 import { promises as fs } from "fs";
```
Comment (fraxken) [inline src/configuration/environment/index.ts:1]: import fs from "fs/promises"

**Pull request merged**